### PR TITLE
Support correlated scalar subqueries (v5)

### PR DESCRIPTION
**Pull request by kokosing**
Support correlated count scalar subqueries

A scalar subquery is a simple query which aggregates all the rows into a
single group (empty GROUP BY clause). For example:

SELECT min(a) FROM t WHERE b > 0
This commit add support for correlated count scalar subqueries which
have WHERE cluase of conjunction of equality comparison. For example:

SELECT (SELECT min(a) FROM t1(a,c) WHERE b = 0 AND c = d) FROM t2(b,d)
Logically such query is rewritten into left outer join where correlation
used in eqality comparisons (from WHERE cluase) are used as join
condition.

### CODE REVIEW

**Comment by kokosing** [comment]
Supersedes #6019

File: presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        }
+
+        Page outputPage = getRowsWithUniqueColumns();
+        inputPage = null;
+        return outputPage;
+    }
+
+    private Page getRowsWithUniqueColumns()
+    {
+        int inputPageChannelCount = inputPage.getChannelCount();
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java]: It should be possible to determine this value "statically" for the operator based on how it's configured at creation time.

File: presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        if (inputPage == null) {
+            return null;
+        }
+
+        Page outputPage = getRowsWithUniqueColumns();
+        inputPage = null;
+        return outputPage;
+    }
+
+    private Page getRowsWithUniqueColumns()
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java]: Maybe call this "processPage"?

File: presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        Block[] outputBlocks = new Block[inputPageChannelCount + 1]; // + 1 for the unique column
+        for (int i = 0; i < inputPageChannelCount; i++) {
+            outputBlocks[i] = inputPage.getBlock(i);
+        }
+        outputBlocks[inputPageChannelCount] = createBlockWithValuesFromPool();
+
+        return new Page(inputPage.getPositionCount(), outputBlocks);
+    }
+
+    private Block createBlockWithValuesFromPool()
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java]: Maybe call this "assignUniqueIds"?

File: presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    {
+        BlockBuilder leastSigBitsBlock = BIGINT.createFixedSizeBlockBuilder(inputPage.getPositionCount());
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java]: Rename this to just "block" or "blockBuilder"

File: presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+        valueCompressor.store(fullTaskId.getStageId().getId(), 10)
+                .store(fullTaskId.getId(), 14);
+
+        requestValues();
+    }
+
+    private void requestValues()
+    {
+        lastCounterValue = valuePool.addAndGet(VALUES_PER_REQUEST);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java]: I think it'd be easier to reason about this if you restructured it like:

```java
counter = valuePool.getAndAdd(VALUES_PER_REQUEST)
lastCounterValue = counter + VALUES_PER_REQUEST;
```

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java]: Also, lastCounterValue is misnamed, since the current batch will not generate a row with that value. Maybe "maxValue"?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java [hunk 6]
```diff
@@ -369,6 +370,12 @@ public StreamProperties visitEnforceSingleRow(EnforceSingleRowNode node, List<St
             return StreamProperties.singleStream();
         }
 
+        @Override
+        public StreamProperties visitAssignUniqueId(AssignUniqueId node, List<StreamProperties> context)
+        {
+            return StreamProperties.singleStream();
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java]: Why "single"? This operator should not affect the scheme of it's input, no?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AssignUniqueId.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class AssignUniqueId
+        extends PlanNode
+{
+    private final PlanNode source;
+    private final Symbol unique;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AssignUniqueId.java]: Call this idColumn or idField

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/PlanNode.java [hunk 8]
```diff
@@ -59,6 +59,7 @@
         @JsonSubTypes.Type(value = GroupIdNode.class, name = "groupid"),
         @JsonSubTypes.Type(value = ExplainAnalyzeNode.class, name = "explainAnalyze"),
         @JsonSubTypes.Type(value = ApplyNode.class, name = "apply"),
+        @JsonSubTypes.Type(value = AssignUniqueId.class, name = "uuid"),
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/PlanNode.java]: UUID is a term of art with specific connotations that don't match what we're doing here. I'd just go with "assignUniqueId".

File: presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java [hunk 9]
```diff
@@ -470,6 +471,15 @@ public Void visitApply(ApplyNode node, Void context)
         }
 
         @Override
+        public Void visitAssignUniqueId(AssignUniqueId node, Void context)
+        {
+            printNode(node, "Unique", NODE_COLORS.get(NodeType.PROJECT));
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java]: AssignUniqueId

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    {
+        return removeFirstRecursive(node);
+    }
+
+    public PlanNode removeFirstRecursive(PlanNode node)
+    {
+        if (where.test(node)) {
+            checkArgument(
+                    node.getSources().size() == 1,
+                    "Unable to remove plan node as it contains many or none children");
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java]: "0 or more than 1 children".

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java [hunk 11]
```diff
@@ -958,5 +959,11 @@ public PlanNode visitTableScan(TableScanNode node, RewriteContext<Expression> co
 
             return node;
         }
+
+        @Override
+        public PlanNode visitAssignUniqueId(AssignUniqueId node, RewriteContext<Expression> context)
+        {
+            return context.defaultRewrite(node, context.get());
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java]: This doesn't look right. What if there's a predicate over the unique id column? It needs to be filtered out at this point.

File: presto-main/src/main/java/com/facebook/presto/sql/ExpressionUtils.java [hunk 12]
```diff
@@ -264,4 +264,14 @@ public Expression rewriteQualifiedNameReference(QualifiedNameReference node, Voi
             }
         }, expression);
     }
+
+    public static Expression andOrTrue(List<Expression> expressions)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/ExpressionUtils.java]: There's a `combineConjuncts` method in this class that does this (and a bit more), already.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 13]
```diff
@@ -0,0 +1,419 @@
...
+import static com.facebook.presto.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Scalar aggregation is aggregation with GROUP BY 1 (or empty GROUP BY). It always returns single row in Presto.
+ * <p>
+ * This optimizer can rewrite correlated scalar aggregation subquery to left outer join in a way described here:
+ * https://docs.google.com/document/d/18HN7peS2eR8lZsErqcmnoWyMEPb6p4OQeidH1JP_EkA/edit#heading=h.hb6ihepq7q74
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: It might be worth dumping this in the Presto wiki to make it more discoverable.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 14]
```diff
@@ -0,0 +1,419 @@
...
+import static com.facebook.presto.sql.planner.optimizations.PlanNodeSearcher.searchFrom;
+import static com.facebook.presto.sql.planner.optimizations.Predicates.isInstanceOfAny;
+import static com.facebook.presto.sql.planner.plan.SimplePlanRewriter.rewriteWith;
+import static com.facebook.presto.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Scalar aggregation is aggregation with GROUP BY 1 (or empty GROUP BY). It always returns single row in Presto.
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: `GROUP BY 1` means "group by first column". I assume you meant something like "group by a constant", no?

**Comment by martint** [comment]
I've merged the first two commits.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java [hunk 15]
```diff
@@ -18,16 +18,23 @@
 import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java]: Add some tests for this change.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 16]
```diff
@@ -0,0 +1,419 @@
...
+            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+        }
+
+        @Override
+        public PlanNode visitApply(ApplyNode node, RewriteContext<PlanNode> context)
+        {
+            ApplyNode rewrittenNode = (ApplyNode) context.defaultRewrite(node, context.get());
+            if (!rewrittenNode.getCorrelation().isEmpty()) {
+                Optional<AggregationNode> scalarAggregation = searchFrom(rewrittenNode.getSubquery())
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: Call this `aggregation`, since it's not necessarily a scalar aggregation.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 17]
```diff
@@ -0,0 +1,419 @@
...
+                        .skipOnlyWhen(isInstanceOfAny(ProjectNode.class, EnforceSingleRowNode.class))
+                        .findFirst();
+                if (scalarAggregation.isPresent() && scalarAggregation.get().getGroupBy().isEmpty()) {
+                    return rewriteScalarAggregation(rewrittenNode, scalarAggregation.get());
+                }
+            }
+            return rewrittenNode;
+        }
+
+        private PlanNode rewriteScalarAggregation(ApplyNode applyNode, AggregationNode scalarAggregation)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: Calling the arguments`apply` and `aggregation`, respectively, may improve readability slightly.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 18]
```diff
@@ -0,0 +1,419 @@
...
+        {
+            List<Symbol> correlation = applyNode.getCorrelation();
+            Optional<RewrittenScalarAggregationSource> rewrittenAggregationSource = rewriteScalarAggregationSource(scalarAggregation.getSource(), correlation);
+            if (!rewrittenAggregationSource.isPresent()) {
+                return applyNode;
+            }
+
+            Symbol nonNull = symbolAllocator.newSymbol("non_null", BooleanType.BOOLEAN);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: What's this? It's not in the example you added to the class javadocs.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 19]
```diff
@@ -0,0 +1,419 @@
...
+                    functions.build(),
+                    scalarAggregation.getMasks(),
+                    ImmutableList.of(groupBySymbols),
+                    scalarAggregation.getStep(),
+                    scalarAggregation.getSampleWeight(),
+                    scalarAggregation.getConfidence(),
+                    scalarAggregation.getHashSymbol()));
+        }
+
+        public Optional<RewrittenScalarAggregationSource> rewriteScalarAggregationSource(PlanNode aggregationSource, List<Symbol> correlation)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: There's nothing specific to this being an "aggregation source", is it? Looking at the code below, this appears to be doing something that can be described as "decorrelating filters". I would just call it "decorrelateFilters" and change the name of the parameter to `node`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 20]
```diff
@@ -0,0 +1,419 @@
...
+                extractConjuncts(((LogicalBinaryExpression) predicate).getLeft(), conjuncts);
+                extractConjuncts(((LogicalBinaryExpression) predicate).getRight(), conjuncts);
+            }
+            else {
+                conjuncts.add(predicate);
+            }
+        }
+    }
+
+    private static class RewrittenScalarAggregationSource
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: Similarly, this class doesn't seem to be specific to "aggregation source", so I'd pick a more generic name.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java (hunk 19 above)
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: Make this method private

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 21]
```diff
@@ -0,0 +1,419 @@
...
+
+        private static boolean isSupportedPredicate(Expression predicate)
+        {
+            AtomicBoolean isSupported = new AtomicBoolean(true);
+            new DefaultTraversalVisitor<Void, AtomicBoolean>()
+            {
+                @Override
+                protected Void visitLogicalBinaryExpression(LogicalBinaryExpression node, AtomicBoolean context)
+                {
+                    if (node.getType() != LogicalBinaryExpression.Type.AND) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: Isn't this too conservative? Should we be able to support expressions that contain ORs and NOTs as long as they are not in any subexpression containing references to correlation columns?

E.g., `x1 = c1 AND x2 = c2 AND ((x3 = 4 OR x4 > x5) AND NOT x6)`, where `c1` and `c2` are the correlation columns and `x1`...`x6` come from the input.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 22]
```diff
@@ -0,0 +1,419 @@
...
+                return Optional.of(new RewrittenScalarAggregationSource(ImmutableList.of(), aggregationSource));
+            }
+
+            Expression predicate = filter.get().getPredicate();
+
+            if (!isSupportedPredicate(predicate)) {
+                return Optional.empty();
+            }
+
+            if (!DependencyExtractor.extractAll(predicate).containsAll(correlation)) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: Since we're looking only at the first filter, the transformation will fail for queries that have multiple filters (say, separated by projections). This may or may not be a problem, depending on where in the sequence of optimizers this one sits.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 23]
```diff
@@ -0,0 +1,419 @@
...
+                return Optional.empty();
+            }
+
+            ImmutableList.Builder<Expression> newPredicate = ImmutableList.builder();
+            ImmutableList.Builder<Expression> joinExpressions = ImmutableList.builder();
+            for (Expression conjunct : extractConjuncts(predicate)) {
+                List<Symbol> symbols = DependencyExtractor.extractAll(conjunct);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: Using `DependencyExtractor.extractUnique`, which returns a `Set`, may be more efficient, since the call `contains` below will be O(1) vs O(N) on the number of symbols in the predicate. This kind of pattern has shown up as a problem in the past for queries with complex expressions or lots of columns.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 24]
```diff
@@ -0,0 +1,419 @@
...
+
+            if (!isSupportedPredicate(predicate)) {
+                return Optional.empty();
+            }
+
+            if (!DependencyExtractor.extractAll(predicate).containsAll(correlation)) {
+                return Optional.empty();
+            }
+
+            ImmutableList.Builder<Expression> newPredicate = ImmutableList.builder();
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: Rather than using these names, I might go with "correlatedPredicates" vs "uncorrelatedPredicates", as that's exactly what the two lists represent.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 25]
```diff
@@ -0,0 +1,419 @@
...
+                }
+            }
+
+            FilterNode newFilterNode = new FilterNode(
+                    idAllocator.getNextId(),
+                    filter.get().getSource(),
+                    ExpressionUtils.andOrTrue(newPredicate.build()));
+
+            if (!joinExpressions.build().isEmpty()) {
+                // filter condition has changed so Limit node no longer applies for EXISTS subquery
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: Can you elaborate on what this comment is referring to?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 26]
```diff
@@ -0,0 +1,419 @@
...
+                List<Symbol> symbols = DependencyExtractor.extractAll(conjunct);
+                if (correlation.stream().anyMatch(symbols::contains)) {
+                    joinExpressions.add(conjunct);
+                }
+                else {
+                    newPredicate.add(conjunct);
+                }
+            }
+
+            FilterNode newFilterNode = new FilterNode(
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: There's no need to create this node if the uncorrelated predicates are empty. You can combine this with the call below that replaces the filter node and do it conditionally.

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: Also, if you move this closer to where the filter node is used, it will make the code easier to read.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 27]
```diff
@@ -0,0 +1,419 @@
...
+                // filter condition has changed so Limit node no longer applies for EXISTS subquery
+                aggregationSource = searchFrom(aggregationSource)
+                        .where(LimitNode.class::isInstance)
+                        .skipOnlyWhen(ProjectNode.class::isInstance)
+                        .removeFirst();
+            }
+            aggregationSource = searchFrom(aggregationSource)
+                    .where(FilterNode.class::isInstance)
+                    .skipOnlyWhen(isInstanceOfAny(ProjectNode.class, LimitNode.class))
+                    .replaceAll(newFilterNode);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: This doesn't look correct. Does it replace *every* filter node recursively? Shouldn't it be just the first node, since that's the one we found at the beginning of this method?

```java
Optional<FilterNode> filter = searchFrom(aggregationSource)
        .where(FilterNode.class::isInstance)
        .skipOnlyWhen(isInstanceOfAny(ProjectNode.class, LimitNode.class))
        .findFirst();
```

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: Incidentally, being so far away from the code that selected the node and without any explanation on how it relates to that code, this seems fragile and prone to someone breaking it accidentally in the future.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 28]
```diff
@@ -0,0 +1,419 @@
...
+
+            aggregationSource = ensureJoinSymbolsAreReturned(aggregationSource, joinExpressions.build());
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: Why is this needed? If "aggregationSource" already returns those columns, there's nothing to be done. If it doesn't, we can't synthesize them from thin air (so that would indicate something is seriously broken in the plan structure).

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 29]
```diff
@@ -0,0 +1,419 @@
...
+                    if (node.getType() != LogicalBinaryExpression.Type.AND) {
+                        context.set(false);
+                    }
+                    return null;
+                }
+            }.process(predicate, isSupported);
+            return isSupported.get();
+        }
+
+        private static List<Expression> extractConjuncts(Expression predicate)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: There's an `extractConjuncts` in ExpressionUtils. How is this one different?

File: presto-product-tests/src/main/resources/sql-tests/testcases/hive_tpch/q1.sql [hunk 30]
```diff
@@ -1,4 +1,4 @@
--- database: presto; groups: tpch, big_query; tables: lineitem
```
Comment (martint) [inline presto-product-tests/src/main/resources/sql-tests/testcases/hive_tpch/q1.sql]: Is this commit independent from adding support for correlated scalar subqueries? If so, I can merge it while you iterate on the other commits.

**Comment by martint** [comment]
Back to you for comments. Getting close!

File: presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java (hunk 5 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java]: done. I also renamed `value` to `rowId`

File: presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java (hunk 2 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java (hunk 1 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java (hunk 3 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java (hunk 4 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java (hunk 6 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java]: I have made it:
```
 return Iterables.getOnlyElement(inputProperties);
```
but I am not sure if it is correct.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AssignUniqueId.java (hunk 7 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AssignUniqueId.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/PlanNode.java (hunk 8 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/PlanNode.java]: In my first approach I used UUID instead of generated unique id. This is a left over after that approach. Fixed.

File: presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java (hunk 9 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java (hunk 10 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java (hunk 11 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java]: I added a `checkState` as currently `uniqueId` column is used only for grouping. I didn't want to filter out this symbol as it would be a dead code without any tests - currently there is no SQL which would produce such case.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java (hunk 15 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java]: It looks like this code is no longer needed. I remember I had to add it at some point where I had some test failures otherwise, but today I removed this code and no failure occurred.

I am going to remove this commit as it looks to be a dead code.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java (hunk 27 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: done

- to make a link between code which find and code which replaces/removes filter node I used the `same planNodeSearcher`
- I use `replaceAll`/`removeAll` as it was checked that there is only a single `FilterNode`. I could change that to `replaceFirst`/`removeFirst` but it would not be coherent with `findAll` at the beginning of this method.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java (hunk 28 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: subquery could be wrapped in a projection which could hide symbols used in `FilterNode`, but these symbols have to be exposed so wrapping `JoinNode` could use them

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java (hunk 21 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: Possibly. Anyway I would prefer to merge conservative version first and then try to extend it to support more queries.

Supporting, more complicated predicates could be tricky.

Lets leave it as it is and then I will take care in separate PR as a separate issue (see https://github.com/prestodb/presto/issues/6154)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java (hunk 29 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: done. I removed this methods. I just didn't know that it was already implemented there.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java (hunk 20 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: I renamed it to `DecorrelatedNode`

**Comment by kokosing** [comment]
@martint All the comments addressed, superseded by #6155

File: presto-product-tests/src/main/resources/sql-tests/testcases/hive_tpch/q1.sql (hunk 30 above)
Comment (kokosing) [inline presto-product-tests/src/main/resources/sql-tests/testcases/hive_tpch/q1.sql]: see https://github.com/prestodb/presto/pull/6156
